test/jdk/tools/sincechecker/SinceChecker.java

```diff
@@ -0,0 +1,828 @@
+    private boolean isDocumented(Element te) {
+        Set<Modifier> mod = te.getModifiers();
+        return mod.contains(Modifier.PUBLIC) || mod.contains(Modifier.PROTECTED);
+    }
+
+    private boolean isMember(Element e) {
+        var kind = e.getKind();
+        return kind.isField() || switch (kind) {
+            case METHOD, CONSTRUCTOR -> true;
+            default -> false;
+        };
+    }
+
+    private void processClassElement(TypeElement te, String version, Types types, Elements elements) {
+        if (!isDocumented(te)) {
+            return;
+        }
+        persistElement(te.getEnclosingElement(), te, types, version);
+        elements.getAllMembers(te).stream()
+                .filter(this::isDocumented)
+                .filter(this::isMember)
+                .forEach(element -> persistElement(te, element, types, version));
+        te.getEnclosedElements().stream()
+                .filter(element -> element.getKind().isDeclaredType())
+                .map(TypeElement.class::cast)
+                .forEach(nestedClass -> processClassElement(nestedClass, version, types, elements));
+    }
+
+    public void persistElement(Element explicitOwner, Element element, Types types, String version) {
+        String uniqueId = getElementName(explicitOwner, element, types);
+        IntroducedIn introduced = classDictionary.computeIfAbsent(uniqueId, i -> new IntroducedIn());
+        if (isPreview(element, uniqueId, version)) {
+            if (introduced.introducedPreview == null) {
+                introduced.introducedPreview = version;
+            }
+        } else {
+            if (introduced.introducedStable == null) {
+                introduced.introducedStable = version;
+            }
+        }
+    }
+
+    private boolean isPreview(Element el, String uniqueId, String currentVersion) {
+        while (el != null) {
+            Symbol s = (Symbol) el;
+            if ((s.flags() & Flags.PREVIEW_API) != 0) {
+                return true;
+            }
+            el = el.getEnclosingElement();
+        }
+
+        return LEGACY_PREVIEW_METHODS.containsKey(currentVersion)
+                && LEGACY_PREVIEW_METHODS.get(currentVersion).contains(uniqueId);
+    }
+
+    void error(String message) {
+        System.err.println(message);
+        errorCount++;
+    }
+
+    private void checkModule(String moduleName) throws Exception {
+        Path home = Paths.get(System.getProperty("java.home"));
+        Path srcZip = home.resolve("lib").resolve("src.zip");
+        if (Files.notExists(srcZip)) {
+            //possibly running over an exploded JDK build, attempt to find a
+            //co-located full JDK image with src.zip:
+            Path testJdk = Paths.get(System.getProperty("test.jdk"));
+            srcZip = testJdk.getParent().resolve("images").resolve("jdk").resolve("lib").resolve("src.zip");
+        }
+        if (!Files.exists(srcZip) && !Files.isDirectory(srcZip)) {
```

Comment:
This condition looks wrong. If `exists` returns false, it also implies that `isDirectory` returns false. I think there's no need to check for a (not-)directory here.

Reply:
Fixed in [3f226ef](https://github.com/openjdk/jdk/pull/18934/commits/3f226ef9134b71a1b63b6562b8381be909c30343), I now only check if the file is readable.